ocaml/src/fragment_cache.ml

```diff
@@ -22,20 +22,31 @@ open Lwt_bytes2
 module KV = Rocks_key_value_store
 
 
-class type cache = object
-    method add : int32 -> string -> Lwt_bytes.t -> unit Lwt.t
-    method lookup : int32 -> string -> Lwt_bytes.t option Lwt.t
-    method lookup2 : int32 -> string -> (int * int * Lwt_bytes.t * int) list -> bool Lwt.t
+class virtual cache ~cache_on_read ~cache_on_write = object(self)
 
-    method drop  : int32 -> global : bool -> unit Lwt.t
-    method close : unit -> unit Lwt.t
+    method virtual _add : int32 -> string -> Bigstring_slice.t -> unit Lwt.t
+    method add bid key blob (moment : [ `Read | `Write ]) =
+      let cache = function
+        | true -> self # _add bid key blob
+        | false -> Lwt.return_unit
+      in
+      match moment with
+      | `Read -> cache cache_on_read
+      | `Write -> cache cache_on_write
```

Comment:
isn't
`if (cache_on_read && moment = `Read) || (cache_on_write && momen = `Write) then ... ` simpler?
Come to think of it. It's a bit odd that the caller knows exactly which path (read or write) she's in and then encodes
it as a parameter for the callee to examine and determine the path from. Maybe it's better to have 2 methods, one for each path.